webapp/src/ts/services/contact-summary.service.ts

```diff
@@ -89,8 +99,10 @@ export class ContactSummaryService {
       uhcInterval: this.uhcStatsService.getUHCInterval(this.visitCountSettings)
     };
 
+    const chtScriptApi = this.chtScriptApiService.getV1Api();
+
     try {
-      const summary = generatorFunction(contact, reports || [], lineage || [], uhcStats, targetDoc);
+      const summary = generatorFunction(contact, reports || [], lineage || [], uhcStats, chtScriptApi, targetDoc);
```

Comment:
this approach is difficult to write automated tests for partner code. how do you recommend people to write unit tests for code which references this api? how can we architect for testability?

Reply:
Can you elaborate on why this would be difficult to test? Firstly do you mean unit tests, integration tests, or both?

I would expect this to be unit testable by mocking out the value for `cht` in the script to do whatever you want. We could develop a stub for this if that would be easier than mocking out the entire API.

For full integration tests then it would use the real logged in user.

For partial integration tests it depends where you're drawing the line.

Reply:
how can we support this in https://github.com/medic/medic-conf-test-harness? specifically, passing angular services into partner code makes that integration test pattern difficult. is it possible to implement the `cht` api in cht-conf so it is compiled in the app bundle? to pass in an npm module (like Utils which already has access to the app_settings) or to abstract this into a shared-lib for re-use in integraiton tests?

Reply:
I like the shared lib idea, because there are parts of this we would want to use in api, sentinel, etc too. Ultimately this should be mocked because the configuration test will be opinionated about the user's roles or permissions in order to assert the correct outcomes. @latin-panda Can you attempt to write or modify a test using the test harness to assert that this design is testable?

The problem with implementing any of this in medic-conf or an npm library like Utils is we want the API to be implementation agnostic, that is to say, if Product decides to change the way permissions are done we want the API to keep working as much as possible. If we include this API in CHT Core then we can guarantee the versions are in sync, but otherwise you get complicated dependencies between Utils, medic-conf, and CHT Core.

> are you already planning to refactor the new uhcStats object into the api prior to 3.12 release?

Unfortunately that work is already ATed, merged, etc so at the moment we'll leave it as is. Once the API design is sorted and we're happy with it (hopefully before 3.12 is released) then we should revisit the uhcStats solution to see if this API is an improvement.

Reply:
Looking at it :)

Reply:
@latin-panda Ping me with any questions about medic-conf-test-harness. Happy to give you [pointers](https://github.com/medic/medic-conf-test-harness/blob/master/all-chts-bundle.js#L28) or fill in your [understanding](https://github.com/medic/medic-conf-test-harness/pull/102#issue-644505398).

Reply:
I think it should be possible adding the cht-api in the test medic-conf-test-harness, but I'm still struggling to run my test in the default config.
`npm ERR! errno ENOTFOUND
npm ERR! network request to https://registry.npmjs.org/cht-core-3-10 failed, reason: getaddrinfo ENOTFOUND registry.npmjs.org
npm ERR! network This is a problem related to network connectivity.
`
I think it's going to be something like this as Kenn suggested: 
Having the cht-api in the shared-libs, but it will need to receive the db documents to do the checks and being able to keep the functions sync for tasks, targets, etc. 

Add the reference:
<img width="642" alt="Screen Shot 2021-06-10 at 7 13 50 pm" src="https://user-images.githubusercontent.com/66472237/121523734-bb980b00-ca20-11eb-82c2-f40e373fcd15.png">

Add the mock data, for example:
<img width="685" alt="Screen Shot 2021-06-10 at 7 15 00 pm" src="https://user-images.githubusercontent.com/66472237/121523713-b8048400-ca20-11eb-879f-211547cb458f.png">

Pass the api: 
<img width="650" alt="Screen Shot 2021-06-10 at 7 14 44 pm" src="https://user-images.githubusercontent.com/66472237/121523731-baff7480-ca20-11eb-9c29-8d2f79e42981.png">

I'll continue with this implementation in cht-core about having the shared-libs tomorrow :)

Reply:
I was able to mock the api in shared-libs and run the medic-conf-test-harness for tasks successfully :) 

The example in the screenshots are about an user with a custom permission "can_make_delivery_form" and another user without that permission. 

<img width="500" alt="Screen Shot 2021-06-11 at 2 03 01 pm" src="https://user-images.githubusercontent.com/66472237/121646076-213ad480-cabf-11eb-86d5-bcacb7d65290.png">

<img width="500" alt="Screen Shot 2021-06-11 at 2 05 55 pm" src="https://user-images.githubusercontent.com/66472237/121646093-26981f00-cabf-11eb-92c2-16fc221167b0.png">